From penberg@cs.helsinki.fi  Tue Apr 21 15:19:59 2009
From: Pekka Enberg <penberg@cs.helsinki.fi>
Date: Thu, 16 Apr 2009 14:43:14 +0300
Subject: Staging: w35und: unregister device from the ieee80211 stack upon ->disconnect()
To: Greg KH <greg@kroah.com>
Cc: luoyi.ly@gmail.com, sandro.bonazzola@gmail.com, pavel@ucw.cz
Message-ID: <1239882194.28965.46.camel@penberg-laptop>


From: Pekka Enberg <penberg@cs.helsinki.fi>

Impact: fix module removal

This patch fixes an oops when the w35und module is removed from the
kernel and added back.

Reported-by: luoyi <luoyi.ly@gmail.com>
Tested-by: Sandro Bonazzola <sandro.bonazzola@gmail.com>
Cc: Pavel Machek <pavel@ucw.cz>
Signed-off-by: Pekka Enberg <penberg@cs.helsinki.fi>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/staging/winbond/wbusb.c |    9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- a/drivers/staging/winbond/wbusb.c
+++ b/drivers/staging/winbond/wbusb.c
@@ -368,7 +368,7 @@ static int wb35_probe(struct usb_interfa
 	if (err)
 		goto error_free_hw;
 
-	usb_set_intfdata(intf, priv);
+	usb_set_intfdata(intf, dev);
 
 	return 0;
 
@@ -397,10 +397,15 @@ static void wb35_hw_halt(struct wbsoft_p
 
 static void wb35_disconnect(struct usb_interface *intf)
 {
-	struct wbsoft_priv *priv = usb_get_intfdata(intf);
+	struct ieee80211_hw *hw = usb_get_intfdata(intf);
+	struct wbsoft_priv *priv = hw->priv;
 
 	wb35_hw_halt(priv);
 
+	ieee80211_stop_queues(hw);
+	ieee80211_unregister_hw(hw);
+	ieee80211_free_hw(hw);
+
 	usb_set_intfdata(intf, NULL);
 	usb_put_dev(interface_to_usbdev(intf));
 }
